umbraco/Umbraco-CMS
# Dynamic Root (Alternative to XPath in MNTP)

### Description
This introduces an alternative to XPath to dynamically find start nodes in Multi Node Tree Picker.

![image](https://github.com/umbraco/Umbraco-CMS/assets/1561480/32890697-3b9b-42b5-91ae-26d8c5deab48)

![image](https://github.com/umbraco/Umbraco-CMS/assets/1561480/83ab5463-be55-45aa-9270-0697051af0b6)

## Details

Its possible to add more query steps using the `DynamicRootQueryStepCollectionBuilder` or add more origins using `DynamicRootOriginCollectionBuilder`.

UI is currently only supported for custom query steps.

Currently all query steps calls a repository that writes a query to find the result.

## Tests
- Verify all origins works
- Verify the implemented steps works as expected
- Verify the custom step works, if you have implemented a custom step.


## Appendix
**Example custom QueryStep**
The following code works if you add a custom step like this:
![image](https://github.com/umbraco/Umbraco-CMS/assets/1561480/a4d4f909-8ab3-45e5-b58a-3f5fd6de237f)

it simply find the roots using code, but this could easily have been used to query something very detailed. e.g. content with property values of "x".

```cs
using System.Diagnostics.CodeAnalysis;
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.DynamicRoot.QuerySteps;
using Umbraco.Cms.Core.Services;

namespace Umbraco.Cms.Web.UI;

public class CustomQueryStepComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.DynamicRootSteps().Append<TestDynamicRootQueryStep>();
    }
}

public class TestDynamicRootQueryStep : IDynamicRootQueryStep
{
    private readonly IContentService _contentService;

    public TestDynamicRootQueryStep(IContentService contentService)
    {
        _contentService = contentService;
    }

    public bool Execute(IEnumerable<Guid> origins, DynamicRootQueryStep filter, [MaybeNullWhen(false)] out IEnumerable<Guid> result)
    {
        if (filter.Alias != "Test" || origins.Any() is false)
        {
            result = null;
            return false;
        }

        result = _contentService.GetRootContent().Select(x => x.Key);

        return true;
    }
}

```
